os/kernel/pthread/pthread_keydelete.c

```diff
@@ -108,5 +110,19 @@
 
 int pthread_key_delete(pthread_key_t key)
 {
-	return ENOSYS;
+	struct pthread_tcb_s *rtcb = (FAR struct pthread_tcb_s *)this_task();
+	struct task_group_s *group = rtcb->cmn.group;
+	int ret = EINVAL;
+
+	DEBUGASSERT(group);
+
+	if (key < PTHREAD_KEYS_MAX && group->tg_keys[key] == IN_USE) {
+		rtcb->pthread_data[key].data = NULL;
+		rtcb->pthread_data[key].destructor = NULL;
+		group->tg_keys[key] = NOT_IN_USE;
+
+		ret = OK;
+	}
+
+	return ret;
```

Comment:
key is not validated for negative value, test case with negative value can over write  the neighbouring data members
How about like below, just  suggestion. you can still take care of key negative in your own way. 
int pthread_key_delete(pthread_key_t key) 
 { 
 	struct pthread_tcb_s *rtcb = (FAR struct pthread_tcb_s *)this_task(); 
 	struct task_group_s *group = rtcb->cmn.group; 
  
 	DEBUGASSERT(group);

	/* Dont allow negative and greater values */ 
	if (key < 0 && key >= PTHREAD_KEYS_MAX)
		return EINVAL;

 	if (group->tg_keys[key] != IN_USE)
		return EINVAL;

	rtcb->pthread_data[key].data = NULL; 
 	rtcb->pthread_data[key].destructor = NULL; 
 	group->tg_keys[key] = NOT_IN_USE; 
 

	return OK;
 }